remote-vfs.c

```diff
@@ -0,0 +1,1889 @@
+ * the drive.  This cache looks like a normal "alternate".
+ *
+ * Officially, the path to the shared cache is stored in the .gvfs folder
+ * (and outside of the repo).  It is available from the "gvfs status" command.
+ * Using it assumes our PATH is properly set up and we know which product's
+ * command to run.
+ */
+static void lookup_odb_paths(void)
+{
+	struct object_directory *odb;
+
+	prepare_alt_odb(the_repository);
+
+	/*
+	 * We now have a linked-list of ODBs.
+	 *
+	 * The first is always the local ODB (inside the .git directory).
+	 * Then come the alternates.  We should only have 1 alternate, but
+	 * can't complain if there are more.
+	 */
+
+	odb = the_repository->objects->odb;
+	if (!odb)
+		return;
+
+	for (odb = odb->next; odb; odb = odb->next) {
+		if (strstr(odb->path, ".gvfsCache"))
+			rv_opts.param__vfs_odb_path = odb->path;
+
+		else if (strstr(odb->path, ".scalarCache"))
+			rv_opts.param__scalar_odb_path = odb->path;
+	}
+}
+
+//////////////////////////////////////////////////////////////////
+
+/*
+ * Select the product-mode from the available inputs.  This resolves
+ * the auto-detect case.  The chosen mode will be used later to select
+ * the various product-specific caches -- or to disable them.
+ */
+static void choose_mode(void)
+{
+	switch (rv_opts.param__mode) {
+	default: /* should not happen, but default to _AUTO. */
+	case RV_MODE_AUTO:
+	{
+		int b_vfs = (rv_opts.param__core_vfs ||
+			     rv_opts.param__vfs_url ||
+			     rv_opts.param__vfs_odb_path);
+		int b_scalar = (rv_opts.param__core_scalar ||
+				rv_opts.param__scalar_url ||
+				rv_opts.param__scalar_odb_path);
+
+		if (b_vfs && b_scalar)
+			die("Both VFS and Scalar found. Cannot auto choose.");
+
+		if (b_vfs)
+			rv_data.chosen_mode = RV_MODE_VFS;
+
+		else if (b_scalar)
+			rv_data.chosen_mode = RV_MODE_SCALAR;
+
+		else
+			rv_data.chosen_mode = RV_MODE_NONE;
+		break;
+	}
+		
+	case RV_MODE_VFS:
+	case RV_MODE_SCALAR:
+	case RV_MODE_NONE:
+		rv_data.chosen_mode = rv_opts.param__mode;
+		break;
+	}
+}
+
+//////////////////////////////////////////////////////////////////
+
+/*
+ * We don't want to allow <user>@ or <user>:<pass>@ forms because
+ * that shortcuts the credential manager (and we want it get us a
+ * PAT), so return a version of the given url without those parts
+ * (and without the complexity of url decoding and re-encoding).
+ *
+ * See "credential_from_url() in credential.c for details.
+ */
+static void strip_auth_from_url(struct strbuf *pbufout, const char *url)
+{
+	const char *at, *colon, *cp, *slash, *host, *proto_end;
+
+	strbuf_setlen(pbufout, 0);
+
+	/*
+	 * Match one of:
+	 *   (1) proto://<host>/...
+	 *   (2) proto://<user>@<host>/...
+	 *   (3) proto://<user>:<pass>@<host>/...
+	 */
+	proto_end = strstr(url, "://");
+	if (!proto_end)
+		return;
+	cp = proto_end + 3;
+	at = strchr(cp, '@');
+	colon = strchr(cp, ':');
+	slash = strchrnul(cp, '/');
+
+	if (!at || slash <= at) {
+		/* Case (1) */
+		host = cp;
+	}
```

Comment:
nit: `} else if (...) {`

Reply:
I'm removing this function and using one that Johannes pointed me to.